### PR TITLE
Fix `_whichmodule` with `multiprocessing`

### LINKED ISSUE #528: `_whichmodule` changes results if  `multiprocessing` is imported first
Without multiprocessing:

```python3

import cloudpickle
from numpy import exp

cloudpickle.cloudpickle._whichmodule(exp, exp.__name__) # numpy.core._multiarray_umath
```

Import `multiprocessing` last:
```python3
import cloudpickle
from numpy import exp
import multiprocessing

cloudpickle.cloudpickle._whichmodule(exp, exp.__name__) # numpy.core._multiarray_umath
```

Import `multiprocessing` first (**problematic case**):
```python3
import multiprocessing
import cloudpickle
from numpy import exp

cloudpickle.cloudpickle._whichmodule(exp, exp.__name__) # __mp_main__
```

Related `pickle` issue: [bpo-42406](https://bugs.python.org/issue42406)

### PR DESCRIPTION
**Pull request by hendrikmakait**
Closes #528

I don't have experience with `cloudpickle`s test suite. Any advice on what test to take as a guideline to test this?

### CODE REVIEW

**Comment by hendrikmakait** [comment]
Thanks, @jrbourbeau! I've added a test now. It seems to require the optional dependency on numpy, at least I haven't been able to reproduce the original issue with the standard library. 

Regarding the workflows, it looks like they require maintainer approval:
> This workflow requires approval from a maintainer. [Learn more about approving workflows.](https://docs.github.com/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks)

**Comment by hendrikmakait** [comment]
@jrbourbeau: Let me know if there's anything else left to do here.

**Review by ogrisel — APPROVED** [review]
I confirm that the non-regression test fails with the previous version of the code and that the fix works.

LGTM, thanks for the fix (I just pushed a cosmetic change and an entry for the changelog).

**Pull request merged**
